src/librustc/middle/const_eval.rs

```diff
@@ -1103,10 +854,28 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
 
           let mut call_args = NodeMap();
           for (arg, arg_expr) in decl.inputs.iter().zip(args.iter()) {
+              let arg_hint = match ty_hint {
+                  ExprTypeChecked => ExprTypeChecked,
+                  UncheckedExprNoHint | UncheckedExprHint(_) => {
+                      if let Some(hint) = tcx.ast_ty_to_ty_cache.borrow().get(&arg.ty.id) {
```

Comment:
@oli-obk 

> Const eval can be called from typeck. In fact this code is only executed before typeck is complete

I see your point, but that very table is itself populated by the same phase in typeck that loads up item signatures, so I don't think this code is really any more robust (also, what @eddyb said:). I think the real sol'n here is probably that const evaluation needs to be better integrated into the astconv process, so that we can handle cycles like this, but I haven't thought deeply about that in some time.